### PR TITLE
Implement guid pool class

### PR DESCRIPTION
**Pull request by Mmduh-483**

### CODE REVIEW

File: pkg/guid/guid_pool.go [hunk 1]
```diff
@@ -15,20 +32,155 @@ type GuidPool interface {
...
 func (p *guidPool) ReleaseGUID(guid string) error {
+	if _, ok := p.guidPoolMap[guid]; !ok {
+		return fmt.Errorf("relase guid %s failed, not allocated ", guid)
+	}
+	delete(p.guidPoolMap, guid)
 	return nil
 }
+
+func parsePodNetworkAnnotation(podNetworks, defaultNamespace string) ([]*multus.NetworkSelectionElement, error) {
+	var networks []*multus.NetworkSelectionElement
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/guid/guid_pool.go]: fixed

File: pkg/guid/guid_pool.go [hunk 2]
```diff
@@ -15,20 +32,155 @@ type GuidPool interface {
...
+
+		return fmt.Errorf("failed to allocate requested guid, already allocated %s", requestedGUID)
+	}
+
+	p.guidPoolMap[requestedGUID] = AllocationStatusAllocated
+
+	return nil
+}
+
+func getNextGUID(currentGUID net.HardwareAddr) net.HardwareAddr {
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/guid/guid_pool.go]: fixed

File: pkg/guid/guid_pool.go [hunk 3]
```diff
@@ -15,20 +32,155 @@ type GuidPool interface {
...
+		}
+
+		copy(p.currentGUID, p.rangeStart)
+	}
+	return "", fmt.Errorf("the range is full")
 }
 
 func (p *guidPool) ReleaseGUID(guid string) error {
+	if _, ok := p.guidPoolMap[guid]; !ok {
+		return fmt.Errorf("relase guid %s failed, not allocated ", guid)
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/guid/guid_pool.go]: fixed

File: pkg/guid/guid_pool.go [hunk 4]
```diff
@@ -15,20 +32,155 @@ type GuidPool interface {
 }
 
 type guidPool struct {
+	kubeClient  k8s_client.Client // kubernetes client
+	rangeStart  net.HardwareAddr  // fist guid in range
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/guid/guid_pool.go]: fixed

File: Gopkg.lock [hunk 5]
```diff
@@ -61,6 +77,18 @@
   revision = "14eae340515388ca95aa8e7b86f0de668e981f54"
   version = "v0.5.4"
 
+[[projects]]
+  branch = "support-guid"
+  digest = "1:ab5db49d9c2900b0bff5d640f494775c7dbc80e02940ec1b88575f8b050d0377"
+  name = "github.com/intel/multus-cni"
+  packages = [
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply Gopkg.lock]: To parse the network annotations which is defined in `multus.types.NetworkSelectionElement`, then extract the existing guid.
This is needed when start the daemon, it obtains all pods and marks all the guids they have as allocated

File: Gopkg.toml [hunk 6]
```diff
@@ -22,6 +22,11 @@
   source = "https://github.com/golang/oauth2.git"
   revision = "bf48bf16ab8d622ce64ec6ce98d2c98f916b6303"
 
+[[override]]
+  name = "github.com/intel/multus-cni"
+  source = "github.com/Mmduh-483/multus-cni.git"
+  branch = "support-guid"
+
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply Gopkg.toml]: It is the branch where guid is support for network annotations

[the remark this answers is not in the document]
Comment (moshe010) [reply Gopkg.toml]: It is not merge remove it

File: pkg/guid/guid_pool.go [hunk 7]
```diff
@@ -15,20 +33,173 @@ type GuidPool interface {
...
+	if _, ok := p.guidPoolMap[guid]; !ok {
+		err := fmt.Errorf("ReleaseGUID(): failed to release guid %s, not allocated ", guid)
+		glog.Error(err)
+		return err
+	}
+	delete(p.guidPoolMap, guid)
 	return nil
 }
+
+func (p *guidPool) parsePodNetworkAnnotation(podNetworks, defaultNamespace string) ([]*multus.NetworkSelectionElement, error) {
```
Comment (moshe010) [inline pkg/guid/guid_pool.go]: Please check if we can reuse the method from multus

File: pkg/guid/guid_pool.go [hunk 8]
```diff
@@ -15,20 +33,173 @@ type GuidPool interface {
...
+
+	p.guidPoolMap[guid] = AllocationStatusAllocated
+
+	return nil
+}
+
+func (p *guidPool) getNextGUID(currentGUID net.HardwareAddr) net.HardwareAddr {
+	for idx := 7; idx >= 0; idx-- {
+		currentGUID[idx] += 1
+		if currentGUID[idx] != 0 {
```
Comment (moshe010) [inline pkg/guid/guid_pool.go]: I don't understand how you make sure that ff:ff:ff:ff:ff:ff:ff:ff will not be generated.
Also it seem you skip if one byte is zero. so valid guid such 11:22:33:44:00:55:66:77 will never be generated

File: pkg/guid/guid_pool.go (hunk 7 above)
Comment (Mmduh-483) [reply pkg/guid/guid_pool.go]: it is private method in multus https://github.com/intel/multus-cni/blob/9874c14e23f52b2bafc8585cfdb96673ae2dad7d/k8sclient/k8sclient.go#L157

Comment (moshe010) [reply pkg/guid/guid_pool.go]: Ok let keep it

File: pkg/guid/guid_pool.go (hunk 8 above)
Comment (Mmduh-483) [reply pkg/guid/guid_pool.go]: the loop start from the least bit, so ff:ff:ff:ff:ff:ff:ff:ff will not generate until the end.

**Comment by Mmduh-483** [comment]
> did you add test case for this? I didn't fine it
I will add one

File: pkg/guid/guid_pool.go [hunk 9]
```diff
@@ -15,20 +32,226 @@ type GuidPool interface {
...
+			break
+		}
+	}
+
+	if strings.ToLower(currentGUID.String()) == "ff:ff:ff:ff:ff:ff:ff:ff" {
```
Comment (moshe010) [inline pkg/guid/guid_pool.go]: instead of compare to ff:ff:ff:ff:ff:ff:ff:ff you need to compare to endrange.

Comment (Mmduh-483) [reply pkg/guid/guid_pool.go]: ok

Comment (Mmduh-483) [reply pkg/guid/guid_pool.go]: the check for last guid then move back to the first guid, is in the `AllocateGUID` function.
This function get the next numeric guid.

Comment (Mmduh-483) [reply pkg/guid/guid_pool.go]: Added test case when last guid in the pool is allocated that the pool will re-check from start again

**Pull request merged**